src/appleseed/renderer/kernel/intersection/triangletree.cpp

```diff
@@ -1504,6 +1515,166 @@ void TriangleLeafVisitor::read_hit_triangle_data() const
     }
 }
 
+bool TriangleLeafVisitor::refine(const ShadingPoint& shading_point)
+{
+    // Compute the location of the intersection point in assembly instance space.
+    ShadingRay::RayType local_ray = shading_point.m_assembly_instance_transform.to_local(shading_point.m_ray);
```

Comment:
We already do that in the code that calls `refine()`, we must absolutely avoid doing it twice (again, performance-critical code).